go/database/mpt/io/interrupt_test.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package io
+
+import (
+	"io"
+	"strings"
+	"sync"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/Fantom-foundation/Carmen/go/database/mpt"
+)
+
+func TestExport_CanBeInterrupted(t *testing.T) {
+	type testFuncs struct {
+		// export is the tested export func
+		export func(string, io.Writer) error
+		// createDB is an init of the database
+		createDB func(t *testing.T, sourceDir string)
+		// check that the interrupted did not corrupt the db by re-opening it
+		check func(t *testing.T, sourceDir string)
+	}
+
+	tests := map[string]testFuncs{
+		"live": {
+			export:   Export,
+			createDB: createTestLive,
+			check:    checkLive,
+		},
+		"archive": {
+			export:   ExportArchive,
+			createDB: createTestArchive,
+			check:    checkArchive,
+		},
+	}
+
+	for name, tf := range tests {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			// Create a small db to be exported.
+			sourceDir := t.TempDir()
+			tf.createDB(t, sourceDir)
+
+			writerWg := new(sync.WaitGroup)
+			writer := &mockWriter{wg: writerWg, signalWrite: false, waitTime: 0}
+			writerWg.Add(1)
+
+			// first find number of writes
+			if err := tf.export(sourceDir, &mockWriter{wg: writerWg}); err != nil {
+				t.Errorf("unexpected error: %v", err)
+				return
+			}
+
+			// save max count and reset number of writes
+			maxCount := writer.numOfWrites
+			writer.signalWrite = true
+			writer.waitTime = 100 * time.Millisecond
+			hasEnded := new(sync.WaitGroup)
+			go func() {
+				hasEnded.Add(1)
+				defer hasEnded.Done()
+				if err := ExportArchive(sourceDir, writer); err != nil {
+					got := err.Error()
+					want := errCanceled.Error()
+					if !strings.Contains(got, want) {
+						t.Errorf("unexpected error: got: %v, want: %v", got, want)
+					}
+					return
+				}
+				t.Error("export was interrupt hence should raise an error")
+			}()
+
+			// wait for first write
+			writerWg.Wait()
+			err := syscall.Kill(syscall.Getpid(), syscall.SIGINT)
+			if err != nil {
+				t.Error("failed to create a SIGINT signal")
+				return
+			}
+
+			// wait for the export to finish
+			hasEnded.Wait()
+			if maxCount == writer.numOfWrites {
```

Comment:
`numOfWrites` must be thread safe - it is read here and modified in the goroutine.